File: apkEditor/src/main/java/com/mcal/apkeditor/dialogs/SmaliCodeDialog.java
package com.mcal.apkeditor.dialogs;

import static com.mcal.common.utils.FileHelperKt.copyFile;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.Dialog;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.res.AssetManager;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.Toast;

import androidx.annotation.NonNull;

import com.mcal.apkeditor.BuildConfig;
import com.mcal.apkeditor.R;

import java.io.Closeable;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.lang.ref.WeakReference;

public class SmaliCodeDialog extends Dialog
        implements android.view.View.OnClickListener {
    static int[] smaliNameIds = {R.string.show_a_toast, R.string.log_a_message,
            R.string.dump_a_value, R.string.print_stack_trace};
    static String[] smaliCodes = {
            "    const-string v0, \"This is a toast.\"\n"
                    + "    # p0 (this object) must be an object of Context\n"
                    + "    invoke-static {p0, v0}, Lapkeditor/Utils;->showToast(Landroid/content/Context;Ljava/lang/String;)V",
            "    # use 'adb logcat APKEDITOR:* *:S' to view the log\n"
                    + "    const-string v0, \"I am here.\"\n"
                    + "    invoke-static {v0}, Lapkeditor/Utils;->log(Ljava/lang/String;)V",
            "    # use 'adb logcat APKEDITOR:* *:S' to view the value\n"
                    + "    invoke-static {v0}, Lapkeditor/Utils;->dumpValue(Ljava/lang/Object;)V",
            "    # use 'adb logcat APKEDITOR:* *:S' to view the stack trace\n"
                    + "    invoke-static {}, Lapkeditor/Utils;->printCallStack()V",
//            "    # 0x60 is the offset, change it to get a different IMEI\n" +
//                    "    const/16 v0, 0x60\n" +
//                    "    invoke-static {v0}, Lapkeditor/Utils;->generateImei(I)Ljava/lang/String;\n" +
//                    "    move-result-object v0",
    };
    private final WeakReference<Activity> activityRef;
    private final String smaliRootFolder;
    private EditText codeEt;

    // filePath: The path for current editing file
    public SmaliCodeDialog(Activity activity, String filePath) {
        super(activity);

        this.activityRef = new WeakReference<>(activity);
        this.smaliRootFolder = getSmaliRootFolder(filePath);

        init(activity);
    }

    @NonNull
    private String getSmaliRootFolder(@NonNull String filePath) {
        String[] dirs = filePath.split("/");
        StringBuilder smaliPath = new StringBuilder();
        for (String dir : dirs) {
            smaliPath.append(dir).append("/");
            if ("smali".equals(dir) || dir.startsWith("smali_")) {
                break;
            }
        }
        return smaliPath.toString();
    }

    @SuppressLint("InflateParams")
    private void init(final Activity activity) {
        View view = LayoutInflater.from(activity).inflate(R.layout.dialog_smalicode, null);

        // Spinner
        Spinner spinner = view.findViewById(R.id.spinner_codename);
        String[] names = new String[smaliNameIds.length];
        for (int i = 0; i < smaliNameIds.length; ++i) {
            names[i] = activity.getString(smaliNameIds[i]);
        }
        ArrayAdapter<String> adapter = new ArrayAdapter<String>(activity, android.R.layout.simple_spinner_item, names);
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        spinner.setAdapter(adapter);

        // Event listener
        spinner.setOnItemSelectedListener(new Spinner.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> arg0, View arg1,
                                       int position, long arg3) {
                updateSmaliCode(position);
            }

            @Override
            public void onNothingSelected(AdapterView<?> arg0) {
            }
        });

        // Code Content
        codeEt = view.findViewById(R.id.et_samplecode);

        // Copy button
        Button copyBtn = view.findViewById(R.id.btn_copy);
        copyBtn.setOnClickListener(this);

        // Close button
        Button closeBtn = view.findViewById(R.id.btn_close);
        closeBtn.setOnClickListener(this);

        setTitle("Smali Code");
        setContentView(view);
    }

    protected void updateSmaliCode(int position) {
        if (position < smaliCodes.length) {
            codeEt.setText(smaliCodes[position]);
        }
    }

    @Override
    public void onClick(@NonNull View v) {
        int id = v.getId();
        if (id == R.id.btn_close) {
            dismiss();
        } else if (id == R.id.btn_copy) {
            // Copy to clipboard
            Activity activity = activityRef.get();
            ClipboardManager clipboard = (ClipboardManager) activity.getSystemService(Context.CLIPBOARD_SERVICE);
            ClipData clip = ClipData.newPlainText("code", codeEt.getText().toString());
            clipboard.setPrimaryClip(clip);

            // Copy Utils.smali to some folder
            copyUtilSmali();

            // Toast
            Toast.makeText(activity, R.string.smali_copied, Toast.LENGTH_SHORT).show();
        }
    }

    // Copy from assets to decode smali folder
    private void copyUtilSmali() {
        String dirPath = this.smaliRootFolder + "apkeditor";
        File dir = new File(dirPath);
        if (!dir.exists()) {
            dir.mkdir();
        }

        String filePath = dirPath + "/Utils.smali";
        File file = new File(filePath);
        if (!file.exists()) {
            FileOutputStream fos = null;
            InputStream is = null;
            try {
                fos = new FileOutputStream(file);
                AssetManager am = activityRef.get().getAssets();
                is = am.open("smali_patch/Utils.smali");
                copyFile(is, fos);
            } catch (Exception e) {
                e.printStackTrace();
            } finally {
                closeQuietly(fos);
                closeQuietly(is);
            }
        }
    }

    private void closeQuietly(Closeable c) {
        if (c != null) {
            try {
                c.close();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }
}

File: feature/code-editor/src/main/java/com/mcal/editor/dialogs/SmaliCodeDialog.kt
package com.mcal.editor.dialogs

import android.app.Activity
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.DialogInterface
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.mcal.common.utils.copyFile
import com.mcal.neweditor.R
import com.mcal.neweditor.databinding.DialogSmaliTemplateBinding
import java.io.*

open class SmaliCodeDialog(private val activity: Activity, filePath: String) {
    private val smaliRootFolder: String
    private val binding = DialogSmaliTemplateBinding.inflate(activity.layoutInflater)
    private val smaliName = arrayOf(
        activity.getString(R.string.show_toast),
        activity.getString(R.string.log_message),
        activity.getString(R.string.dump_value),
        activity.getString(R.string.print_stacktrace)
    )

    init {
        smaliRootFolder = getSmaliRootFolder(filePath)

        val names = arrayOfNulls<String>(smaliName.size)
        System.arraycopy(smaliName, 0, names, 0, smaliName.size)

        binding.spinnerCodename.apply {
            adapter = ArrayAdapter(activity, android.R.layout.simple_spinner_item, names).apply {
                setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
            }
            onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
                override fun onItemSelected(arg0: AdapterView<*>?, arg1: View, position: Int, arg3: Long) {
                    updateSmaliCode(position)
                }

                override fun onNothingSelected(arg0: AdapterView<*>?) {}
            }
        }

        val dialog = MaterialAlertDialogBuilder(activity)
            .setView(binding.root)
            .setPositiveButton(activity.getString(R.string.copy), null)
            .setNegativeButton(android.R.string.cancel, null)
            .create()

        dialog.setOnShowListener {
            dialog.getButton(DialogInterface.BUTTON_POSITIVE)?.setOnClickListener {
                val clipboard = activity.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
                val clip = ClipData.newPlainText("code", binding.etSamplecode.text.toString())
                clipboard.setPrimaryClip(clip)
                copyUtilSmali()
            }
            dialog.getButton(DialogInterface.BUTTON_NEGATIVE)?.setOnClickListener {
                dialog.dismiss()
            }
        }
        dialog.show()
    }

    private fun getSmaliRootFolder(filePath: String): String {
        val dirs = filePath.split("/".toRegex()).toTypedArray()
        val smaliPath = StringBuilder()
        for (dir in dirs) {
            smaliPath.append(dir).append("/")
            if ("smali" == dir || dir.startsWith("smali_")) {
                break
            }
        }
        return smaliPath.toString()
    }

    protected fun updateSmaliCode(position: Int) {
        if (position < smaliCodes.size) {
            binding.etSamplecode.setText(smaliCodes[position])
        }
    }

    // Copy from assets to decode smali folder
    private fun copyUtilSmali() {
        val dirPath = smaliRootFolder + "apkeditor"
        val dir = File(dirPath)
        if (!dir.exists()) {
            dir.mkdir()
        }
        val filePath = "$dirPath/Utils.smali"
        val file = File(filePath)
        if (!file.exists()) {
            var fos: FileOutputStream? = null
            var inputStream: InputStream? = null
            try {
                fos = FileOutputStream(file)
                inputStream = activity.assets.open("smali_patch/Utils.smali")
                copyFile(inputStream, fos)
            } catch (e: Exception) {
                e.printStackTrace()
            } finally {
                closeQuietly(fos)
                closeQuietly(inputStream)
            }
        }
    }

    private fun closeQuietly(c: Closeable?) {
        try {
            c?.close()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    companion object {
        private val smaliCodes = arrayOf(
            """    const-string v0, "This is a toast."
    # p0 (this object) must be an object of Context
    invoke-static {p0, v0}, Lapkeditor/Utils;->showToast(Landroid/content/Context;Ljava/lang/String;)V""",
            """    # use 'adb logcat APKEDITOR:* *:S' to view the log
    const-string v0, "I am here."
    invoke-static {v0}, Lapkeditor/Utils;->log(Ljava/lang/String;)V""",
            """    # use 'adb logcat APKEDITOR:* *:S' to view the value
    invoke-static {v0}, Lapkeditor/Utils;->dumpValue(Ljava/lang/Object;)V""",
            """    # use 'adb logcat APKEDITOR:* *:S' to view the stack trace
    invoke-static {}, Lapkeditor/Utils;->printCallStack()V"""
        )
    }
}